### PR TITLE
[GeoMechanicsApplication] Add 2D rotation matrix calculation for interfaces

### PR DESCRIPTION
**Pull request by rfaasse**
**📝 Description**
This PR adds a utility that calculates the rotation matrix given a certain local coordinate $\xi$, given a line interface geometry. It also includes splitting the interface element utilities in a .h and .cpp. A lot of the other functions in this class can be removed in time, because we now also have an automated way to create the same 'Nu matrix'.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+void InterfaceElementUtilities::CalculateLinkPermeabilityMatrix(BoundedMatrix<double, 3, 3>& rPermeabilityMatrix,
+                                                                const double& JointWidth)
+{
+    // Prism_interface_3d_6 and Hexahedral_interface_3d_8
+    rPermeabilityMatrix(0, 0) = JointWidth * JointWidth / 12.0;
+    rPermeabilityMatrix(1, 1) = JointWidth * JointWidth / 12.0;
+    rPermeabilityMatrix(2, 2) = JointWidth * JointWidth / 12.0;
+}
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp:175]: These functions are just moved. I agree they should be improved and actually most of them can be removed, after we have our interface stress state, but that is out of scope for this PR in my opinion

Comment (WPK4FEM) [reply applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp:175]: Indeed, don't put other efforts here as this is supposed to disappear.

File: applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    std::transform(rGeometry.begin(), half_way_point, half_way_point, std::back_inserter(result),
+                   [](const Node& rNode1, const Node& rNode2) -> array_1d<double, 3> {
+        return 0.5 * (rNode1 + rNode2);
+    });
+
+    return result;
+}
+
+array_1d<double, 3> InterfaceElementUtilities::CalculateTangentialVector(
+    const Matrix& rShapeFunctionGradients, const std::vector<array_1d<double, 3>>& rLocations)
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp]: Locations is quite general. Probably NodalCoordinates is more descriptive as other locations should not be used.

File: applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    // clang-format on
+
+    return result;
+}
+
+std::vector<array_1d<double, 3>> InterfaceElementUtilities::CalculateMidPoints(const Geometry<Node>& rGeometry)
+{
+    std::vector<array_1d<double, 3>> result;
+
+    const auto half_way_point = rGeometry.begin() + rGeometry.PointsNumber() / 2;
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp]: half_way_point is confusing for me. In the interface this is the first point on the opposite side. We should find some common way of describing the interface sides to which we stick forever after.

File: applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+array_1d<double, 3> InterfaceElementUtilities::CalculateTangentialVector(
+    const Matrix& rShapeFunctionGradients, const std::vector<array_1d<double, 3>>& rLocations)
+{
+    array_1d<double, 3> tangential_vector = ZeroVector(3);
+    for (std::size_t i = 0; i < rLocations.size(); ++i) {
+        tangential_vector += rLocations[i] * rShapeFunctionGradients(i, 0);
+    }
+    tangential_vector /= norm_2(tangential_vector);
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/custom_utilities/interface_element_utilities.cpp]: I wonder if we can write this down as a vector operation. What seems to block us is the type array_1d i.s.o. something like a ublas vector.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    nodes.push_back(Kratos::make_intrusive<Node>(2, 5.0, 0.0, 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(3, 0.0, 0.0, 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(4, 5.0, 0.0, 0.0));
+    const LineInterfaceGeometry geometry(1, nodes);
+
+    // Act
+    const auto local_coordinate = array_1d<double, 3>{0.0, 0.0, 0.0};
+    Matrix rotation_matrix = InterfaceElementUtilities::Calculate2DRotationMatrix(geometry, local_coordinate);
+
+    // Assert
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp]: This looks like comment for the sake of comment to me. It repeats throughout and KRATOS_EXPECT_ is quite self explaining.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    Matrix expected_rotation_matrix(2, 2);
+    expected_rotation_matrix <<= 0.5 * sqrt(3), 0.5, -0.5, 0.5 * sqrt(3); // Rotation of 30 degrees clockwise
+    KRATOS_INFO("Expected rotation matrix") << expected_rotation_matrix << std::endl;
+    KRATOS_EXPECT_MATRIX_NEAR(expected_rotation_matrix, rotation_matrix, 1e-6)
+    KRATOS_INFO("End test") << std::endl;
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp]: Left behind debug stuff KRATOS_INFO.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+{
+    PointerVector<Node> nodes;
+    nodes.push_back(Kratos::make_intrusive<Node>(1, 0.0, 0.0, 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(2, 5.0, 0.0, 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(3, 0.0, 0.0, 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(4, 5.0, 0.0, 0.0));
+    const LineInterfaceGeometry geometry(1, nodes);
+
+    // Act
+    const auto local_coordinate = array_1d<double, 3>{0.0, 0.0, 0.0};
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp:33]: If anything needs comment here, it is that the local_coordinate has 2 extra unused components.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    PointerVector<Node> nodes;
+    nodes.push_back(Kratos::make_intrusive<Node>(1, -0.5, 0.5 * std::sqrt(3), 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(2, 0.0, 0.0, 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(3, -0.5, 0.5 * std::sqrt(3), 0.0));
+    nodes.push_back(Kratos::make_intrusive<Node>(4, 0.0, 0.0, 0.0));
+    const LineInterfaceGeometry geometry(1, nodes);
+
+    // Act
+    // Since the gradient of the shape functions is constant, the rotation matrix is the same at
+    // point, meaning the local_coordinate should not have an effect
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp]: same at each point? missing word.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    nodes.push_back(Kratos::make_intrusive<Node>(4, 0.5 * std::sqrt(3), -0.5, 0.0));
+    const LineInterfaceGeometry geometry(1, nodes);
+
+    // Act
+    const auto local_coordinate = array_1d<double, 3>{0.0, 0.0, 0.0};
+    Matrix rotation_matrix = InterfaceElementUtilities::Calculate2DRotationMatrix(geometry, local_coordinate);
+
+    // Assert
+    Matrix expected_rotation_matrix(2, 2);
+    expected_rotation_matrix <<= 0.5 * sqrt(3), 0.5, -0.5, 0.5 * sqrt(3); // Rotation of 30 degrees clockwise
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp]: Using layout to let the values appear as a matrix helps me a lot to see structure. Same holds further down.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+    // Act
+    const auto local_coordinate = array_1d<double, 3>{0.0, 0.0, 0.0};
+    Matrix rotation_matrix = InterfaceElementUtilities::Calculate2DRotationMatrix(geometry, local_coordinate);
+
+    // Assert
+    // clang-format off
+    Matrix expected_rotation_matrix(2, 2);
+    expected_rotation_matrix <<= 0.5 * sqrt(2), 0.5 * sqrt(2),
+                                -0.5 * sqrt(2), 0.5 * sqrt(2); // Rotation of 45 degrees clockwise
```
Comment (WPK4FEM) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp:179]: Nice, helpful layout.

**Review by WPK4FEM — COMMENTED** [review]
Dear Richard,
Very extensively covered new functionality. Most of my comments are trivial, but some of the naming can really be improved.
Thank you,
Wijtze Pieter

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp (hunk 5 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_interface_element_utilities.cpp]: It's there to make the 'Arrange, Act, Assert' pattern clear in the unit test. However, I'm also fine with removing it, what is your opinion here @avdg81?

**Comment by rfaasse** [comment]
Closing this PR, as we created a clean one at #12663
